Spark: Support building against both Spark 3.0 and Spark 3.1.

wypoon opened the pull request:
Code changes that allow spark3 and spark3-extensions to be built against both Spark 3.0 and Spark 3.1.
- A method from `org.apache.spark.sql.catalyst.util.DateTimeUtils` that has changed its name is copied to `org.apache.iceberg.util.DateTimeUtil` and the new method is used instead.
- The trait, `org.apache.spark.sql.catalyst.plans.logical.V2WriteCommand`, has 3 additional methods that need to be implemented. They are implemented in `ReplaceData` but without `override`.
- The trait, `org.apache.spark.sql.catalyst.parser.ParserInterface`, no longer has the `parseRawDataType` method, so `IcebergSparkSqlExtensionsParser` implements it without `override`, by simply throwing `UnsupportedOperationException`, as the method is not used in Iceberg.
- The main constructor for `org.apache.spark.sql.catalyst.expressions.SortOrder` has changed its signature. Use reflection to create instances of `SortOrder`.
- The constructor for `org.apache.spark.sql.execution.datasources.v2.DataSourceV2ScanRelation` has changed its signature. Use reflection to create instances of `DataSourceV2ScanRelation`.
- `org.apache.spark.sql.catalyst.SQLConfHelper` was introduced in Spark 3.1 and a number of classes and traits now extend it, or, in the case of `org.apache.spark.sql.catalyst.analysis.CastSupport`, has it as a self type. This is the trickiest part. I move the mixin, `CastSupport` from `AssignmentAlignmentSupport` to the rule, `AlignRowLevelOperations`, since `org.apache.spark.sql.catalyst.rules.Rule` implements  `SQLConfHelper`. I define the `conf` method in the traits `AssignmentAlignmentSupport` and `RewriteRowLevelOperationHelper`, so that it can be overridden in the classes that extend them, which also extend `Rule[LogicalPlan]` (and thus `SQLConfHelper` in Spark 3.1). When compiling with Spark 3.0, the `conf` in the Iceberg traits are overridden, and when compiling with Spark 3.1, the `conf` in `SQLConfHelper` is overridden.
- The constructor for `org.apache.spark.sql.catalyst.expressions.Alias` has changed its signature to add another parameter to its second parameter list. Use reflection to create instances of `Alias`. This issue is only encountered at runtime when using a different version of Spark than built against.
- The constructor for `org.apache.spark.sql.catalyst.plans.logical.RepartitionByExpression` has changed its signature. Use reflection to create instances of `RepartitionByExpression`. This issue is only encountered at runtime when using a different version of Spark than built against.

On the build side:
- spark3 and spark3-extensions continue to build against Spark 3.0. New tasks are added to build the tests (using main classes built against Spark 3.0) and run them against Spark 3.1.
- spark3-runtime builds against Spark 3.0 (since it depends on spark3 and spark3-extensions that build against 3.0), but a new integration test task is added to build and run the integration tests against Spark 3.1 using the runtime jar (built against 3.0).

wypoon commented on spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/ReplaceData.scala:
```diff
@@ -25,4 +25,12 @@ import org.apache.spark.sql.connector.write.BatchWrite
-    query: LogicalPlan) extends V2WriteCommand
+    query: LogicalPlan) extends V2WriteCommand {
+
+  def isByName: Boolean = false
```
I am not sure if we want this to be true or false. I suspect it doesn't really matter.

rdblue replied:
It shouldn't matter.

wypoon replied:
Ok, thanks.

wypoon commented on spark3/src/test/java/org/apache/iceberg/spark/sql/TestDeleteFrom.java:
```diff
@@ -41,6 +43,7 @@ public void removeTables() {
 
   @Test
   public void testDeleteFromUnpartitionedTable() {
+    Assume.assumeFalse(Spark3VersionUtil.isSpark31());
```
@aokolnychyi this test and the next fail in Spark 3.1. In https://github.com/apache/spark/blob/v3.1.1/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala#L250-L253, if `canDeleteWhere` returns false, there is no attempt to rewrite the query, an exception is just thrown. Is Spark supposed to rewrite the query if `SupportsDelete#canDeleteWhere` returns false? If so, this part hasn't been implemented in Spark yet.

wypoon commented:
@RussellSpitzer @rdblue can you please review and suggest what to do on the build side? I am not very familiar with gradle and am also not sure how we want to do this.

RussellSpitzer commented on spark3/src/main/java/org/apache/iceberg/spark/Spark3VersionUtil.java:
```diff
@@ -0,0 +1,42 @@
+
+package org.apache.iceberg.spark;
+
+public class Spark3VersionUtil {
```
I think we can just check the build constant here, no need to do reflection for this,

scala> org.apache.spark.SPARK_VERSION_SHORT
res10: String = 3.0.2

wypoon replied:
I adopted your suggestion.

RussellSpitzer commented:
Took a brief look but I think once the gradle fixes are in there will be more changes to make so I didn't go over the whole thing.

I think what we'll need to do is make a new configuration for Spark3.1 Testing (and maybe a new integration configuration for Spark3-Runtime)

something like

```
  configurations {
     spark30test extends from testRuntime
     spark31test extends from testRuntime
   }
```
Then in dependencies add the Spark dependency you want to each of those configurations.

That's at least how I would do it

wypoon commented:
@RussellSpitzer thanks for taking a look and the pointer. It doesn't make immediate sense to me (that's how little I know about gradle), but I'll do some research.

wypoon commented:
CI is green.
Building spark3 and spark3-extensions against both 3.0 and 3.1 ensures that the code can compile against both. spark3-runtime can be built against either, up to the user; it builds against 3.0 by default. I think this is a good compromise.
In principle, we can also build against 3.0 and run (test) against 3.1, but I think this approach is a good tradeoff and would rather not add one more combination to the CI.

RussellSpitzer commented:
I will definitely take a look soon.

rdblue commented on spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSparkSqlExtensionsParser.scala:
```diff
@@ -41,7 +43,17 @@ import org.apache.spark.sql.types.StructType
+      .impl(classOf[VariableSubstitution])
+      .impl(classOf[VariableSubstitution], classOf[SQLConf])
+      .build()
+  private lazy val substitutor =
```
The constructor should automatically drop extra arguments, so you should only need `substitutorCtor.newInstance(SQLConf.get)`

rdblue commented on spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSparkSqlExtensionsParser.scala:
```diff
@@ -54,8 +66,13 @@ class IcebergSparkSqlExtensionsParser(delegate: ParserInterface) extends ParserI
    */
-  override def parseRawDataType(sqlText: String): DataType = {
-    delegate.parseRawDataType(sqlText)
+  def parseRawDataType(sqlText: String): DataType = {
```
I don't think that Iceberg actually uses this method anywhere, so I would opt for a simpler approach: either throw `UnsupportedOperationException` or return `parseDataType(sqlText)`.

wypoon replied:
In that case, I'll just throw `UnsupportedOperationException` here.

rdblue commented on spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/DistributionAndOrderingUtils.scala:
```diff
@@ -118,7 +120,24 @@ object DistributionAndOrderingUtils {
+              classOf[Set[catalyst.expressions.Expression]])
+            .build()
+        if (Spark3VersionUtil.isSpark30) {
+          ctor.newInstance(catalystChild, toCatalyst(s.direction), toCatalyst(s.nullOrdering), Set.empty)
```
Does `Set` implement `Seq`? If so, does it work to just use the `Set.empty` variation here?

wypoon replied:
`Set` does not extend `Seq` nor does `Seq` extend `Set`. I don't know of any common subclass of both in the standard Scala library.

rdblue commented on spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/RewriteRowLevelOperationHelper.scala:
```diff
@@ -229,4 +232,26 @@ object RewriteRowLevelOperationHelper {
+      relation: DataSourceV2Relation,
+      scan: Scan,
+      outputAttrs: Seq[AttributeReference]): DataSourceV2ScanRelation = {
+    val ctor: DynConstructors.Ctor[DataSourceV2ScanRelation] =
```
This constructor should be moved into a companion object so that the lookups don't happen every time the method is called.

wypoon replied:
The `createScanRelation` method is already in the companion object. I moved the val defining the constructor to the object too, so it is not created each time the method is called.

rdblue commented on spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/DistributionAndOrderingUtils.scala:
```diff
@@ -118,7 +120,24 @@ object DistributionAndOrderingUtils {
       case s: SortOrder =>
         val catalystChild = toCatalyst(s.expression(), query, resolver)
-        catalyst.expressions.SortOrder(catalystChild, toCatalyst(s.direction), toCatalyst(s.nullOrdering), Set.empty)
+        val ctor: DynConstructors.Ctor[catalyst.expressions.SortOrder] =
```
This constructor should be moved into a companion object so that the lookups don't happen every time the method is called.

wypoon replied:
Moved the val defining the constructor from a local val to a val in the object.

rdblue commented on spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSparkSqlExtensionsParser.scala:
```diff
@@ -41,7 +43,17 @@ import org.apache.spark.sql.types.StructType
 class IcebergSparkSqlExtensionsParser(delegate: ParserInterface) extends ParserInterface {
 
-  private lazy val substitutor = new VariableSubstitution(SQLConf.get)
+  private lazy val substitutorCtor: DynConstructors.Ctor[VariableSubstitution] =
```
Can you move this to a companion object so it doesn't happen for every instance of the extensions?

wypoon replied:
Done.

rdblue commented on spark3/src/test/java/org/apache/iceberg/spark/sql/TestDeleteFrom.java:
```diff
@@ -41,6 +43,7 @@ public void removeTables() {
 
   @Test
   public void testDeleteFromUnpartitionedTable() {
+    Assume.assumeTrue(Spark3VersionUtil.isSpark30());
```
Can you please add context when disabling tests to explain why Spark 3.0 is needed?

wypoon replied:
I'll add an explanatory comment.

rdblue commented on build.gradle:
```diff
@@ -921,33 +921,77 @@ if (jdkVersion == '8') {
+  }
+
+  dependencies project.ext.spark30Dependencies
+  dependencies project.ext.commonDependencies
```
Why refactor common dependencies like this as well?

rdblue commented on build.gradle:
```diff
@@ -921,33 +921,77 @@ if (jdkVersion == '8') {
+  }
+}
+
+project(':iceberg-spark31') {
```
Do the changes work when using the compiled artifact from `iceberg-spark3` but testing in Spark 3.1? That's what we're ideally trying to achieve. It is nice to share source, but we would prefer not to have a separate module for every Spark version.

rdblue commented on build.gradle:
```diff
@@ -1031,15 +1129,29 @@ project(':iceberg-spark3-runtime') {
   }
 
+  ext {
+    prop = findProperty('spark31Runtime')
```
If we need to have a separate module, then we should not enable it with a property. We should just add another runtime module. Otherwise, it gets very confusing that you can produce two artifacts with different compatibility from the same module.

wypoon replied:
I removed the use of the property and added another runtime module instead.

RussellSpitzer commented:
I think I would prefer 2 test configurations like I suggested before. That way we run the suite against both spark versions with the same compiled code. I can try throwing together an example if we get some of the Compaction stuff done this week.

RussellSpitzer commented:
@wypoon + @rdblue  Wrote up a little example about we can do a similar multi-test version here
https://github.com/RussellSpitzer/iceberg/commit/7ee33d90cb3299fb90c7dd4707e16078eee89904

Basically what we do in this example is have 2 test tasks
The default which uses 3.0.2
And testSpark31 which uses 3.1

In both cases the test code itself is compiled against the respective Spark version but the library code is always based on the 3.0.2 code. So basically we make sure our library classes work regardless of which Spark version you use them with. I only implemented the Spark3 config for this but we could duplicate it for Extensions and Integration tests pretty easily. Let me know what you think.

RussellSpitzer commented:
@aokolnychyi I don't do releases so you should probably double check :)

wypoon commented:
@RussellSpitzer I tried out your build.gradle. Unfortunately, it appears that an Iceberg Spark3 runtime jar built against Spark 3.0 does not completely work with Spark 3.1. The integration test, `SmokeTest` fails in `testGettingStarted` when running a `MERGE INTO`, due to incompatible signatures in the Spark case class `Alias`:
```
java.lang.NoSuchMethodError: org.apache.spark.sql.catalyst.expressions.Alias.<init>(Lorg/apache/spark/sql/catalyst/expressions/Expression;Ljava/lang/String;Lorg/apache/spark/sql/catalyst/expressions/ExprId;Lscala/collection/Seq;Lscala/Option;)V
	at org.apache.spark.sql.catalyst.optimizer.RewriteMergeInto$$anonfun$apply$1.applyOrElse(RewriteMergeInto.scala:156)
```
This is because the case class added an additional parameter in 3.1 vs 3.0:
https://github.com/apache/spark/blob/branch-3.0/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala#L146-L150
https://github.com/apache/spark/blob/branch-3.1/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala#L149-L154

@rdblue I'm going to stick with my approach, introduce new modules to build and test spark3 and spark3-extensions against Spark 3.1, and also to introduce a new runtime module for Spark 3.1. We will produce two Spark3 runtime jars explicitly, one for 3.0 and one for 3.1. I won't use a build property.

wypoon commented:
Excerpt from my gradle output:
```
> Task :iceberg-spark3-runtime:integrationTest
[Thread-8] INFO org.apache.spark.util.ShutdownHookManager - Shutdown hook calle
[Thread-8] INFO org.apache.spark.util.ShutdownHookManager - Deleting directory /private/var/folders/7x/vgfk2h155v9gjtw55z6ncgc00000gp/T/spark-51d1d3aa-5b0a-48fa-8dec-70ef4725e0be

> Task :iceberg-spark3-runtime:spark31IntegrationTest

org.apache.iceberg.spark.SmokeTest > testGettingStarted[catalogName = testhive, implementation = org.apache.iceberg.spark.SparkCatalog, config = {type=hive, default-namespace=default}] FAILED
    java.lang.NoSuchMethodError at SmokeTest.java:67

org.apache.iceberg.spark.SmokeTest > testGettingStarted[catalogName = testhadoop, implementation = org.apache.iceberg.spark.SparkCatalog, config = {type=hadoop}] FAILED
    java.lang.NoSuchMethodError at SmokeTest.java:67

org.apache.iceberg.spark.SmokeTest > testGettingStarted[catalogName = spark_catalog, implementation = org.apache.iceberg.spark.SparkSessionCatalog, config = {type=hive, default-namespace=default, parquet-enabled=true, cache-enabled=false}]FAILED
    java.lang.NoSuchMethodError at SmokeTest.java:67
[shutdown-hook-0] INFO org.apache.spark.util.ShutdownHookManager - Shutdown hook called
[shutdown-hook-0] INFO org.apache.spark.util.ShutdownHookManager - Deleting directory /private/var/folders/7x/vgfk2h155v9gjtw55z6ncgc00000gp/T/spark-93a9a974-bcaf-4d43-99c8-2b4b97f6a9ed

9 tests completed, 3 failed

> Task :iceberg-spark3-runtime:spark31IntegrationTest FAILED

FAILURE: Build failed with an exception.

```
(@RussellSpitzer I renamed your testSpark31 task to spark31IntegrationTest. As you can see integrationTest passed, but spark31IntegrationTest failed.)

wypoon commented:
@RussellSpitzer and @rdblue, can you please review this again?

rdblue commented:
@wypoon, it's much easier for users if we can use shims to handle API compatibility issues. Is Alias the only problem? Can we use a Spark shim to instantiate it?

wypoon commented:
> @wypoon, it's much easier for users if we can use shims to handle API compatibility issues. Is Alias the only problem? Can we use a Spark shim to instantiate it?

I can handle Alias by using reflection (DynConstructors) as well, but the real issue is this: other compatibility differences that are currently addressed are all found at compile-time; the problem with Alias was found at runtime, and only because there is an integration test that exercised that code path. The integration test suite is just a smoke test. There is no telling what other compatibility issues there are. That is why my preference is to build separate runtime jars.

RussellSpitzer commented:
I think that means we should finish up the pr I started which runs the old code with test code compiled against the new spark.

wypoon commented:
> I think that means we should finish up the pr I started which runs the old code with test code compiled against the new spark.

Actually, I did try out your change. However, the unit tests in spark3 and spark3-extensions did not catch any issues. It was the integration test in spark3-runtime using the runtime jar built against 3.0 with Spark 3.1 that encountered the problem with Alias. (In that change, the integration test is run with Spark 3.0 and with Spark 3.1; there is only one runtime jar, built against 3.0.)

RussellSpitzer commented:
That just means I must have done something wrong :) And we should fix it.

wypoon commented:
My overall point is that a runtime jar built with Spark 3.1 will not have compatibility issues with 3.1. One built with 3.0 may or may not have compatibility issues; finding issues is a function of how comprehensive the tests are. As new functionality gets written, the potential for compatibility problems will increase.

wypoon commented:
@RussellSpitzer I have changed the build to build spark3 and spark3-extensions against Spark 3.0, and added tasks to build and run the tests in those two modules against Spark 3.1. spark3-runtime is thus also built against Spark 3.0,  but a task is added that builds and runs the integration test against Spark 3.1 using the runtime jar (built against 3.0). The build.gradle is based on the one in your PR, but I had to make some modifications in order to get the Spark 3.1 tests in spark3 and spark3-extensions to actually run (before that, they were not getting run, so we were only testing Spark 3.0).
@rdblue please take a look and see if you are fine with this approach. I have made additional changes to the code to account for runtime invocations that were not caught before at compile-time. As I have stated, there could be more incompatibilities that we have not caught because our tests don't exercise some code paths.
If you are fine with this, I'll rebase on master and make sure that changes made to spark3 and spark3-extensions since I started this PR are accounted for as well.

wypoon commented:
Ok, I rebased on master. There don't seem to be any additional changes necessary to get a successful build.

RussellSpitzer commented:
@wypoon Do we need a testSpark31 for spark3-extensions too? Or do we have no code split there?

RussellSpitzer commented:
I think this looks pretty good outside of the spark3-extensions test case :+1:

wypoon commented on build.gradle:
```diff
@@ -1001,6 +1048,15 @@ project(":iceberg-spark3-extensions") {
     arguments += ['-visitor', '-package', 'org.apache.spark.sql.catalyst.parser.extensions']
   }
+
+  task testSpark31(type: Test) {
+    dependsOn classes
+    description = "Test against Spark 3.1"
+    testClassesDirs = sourceSets.spark31.output.classesDirs
+    classpath = sourceSets.spark31.runtimeClasspath + sourceSets.main.output
+  }
+
+  test.dependsOn testSpark31
```
@RussellSpitzer this is the testSpark31 for spark3-extensions.

wypoon commented:
> @wypoon Do we need a testSpark31 for spark3-extensions too? Or do we have no code split there?

There is a testSpark31 for spark3 and spark3-extensions (you must have missed it somehow). For spark3-runtime, a spark31IntegrationTest is added.

RussellSpitzer commented:
Excellent! I missed that when I was looking through the test dependencies

On Mon, Jun 7, 2021 at 3:59 PM Wing Yew Poon ***@***.***>
wrote:

> @wypoon <https://github.com/wypoon> Do we need a testSpark31 for
> spark3-extensions too? Or do we have no code split there?
>
> There is a testSpark31 for spark3 and spark3-extensions (you must have
> missed it somehow). For spark3-runtime, a spark31IntegrationTest is added.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/iceberg/pull/2512#issuecomment-856252963>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADE2YI2JALLQZA2TZZASM3TRUXJRANCNFSM43R5UBZQ>
> .
>

wypoon commented:
Updated description to reflect latest changes.

wypoon commented:
@RussellSpitzer are you ok with this change? can you merge it? @rdblue do you want to review it again?
As new changes go into spark3 and spark3-extensions, this change may need to be rebased and retested. A timely review is very much appreciated! Thank you!

wypoon commented:
I rebased on master and resolved the conflicts. CI is green.
@RussellSpitzer @rdblue can you help move this forward?

pan3793 commented on spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/DistributionAndOrderingUtils.scala:
```diff
@@ -80,7 +94,11 @@ object DistributionAndOrderingUtils {
       // this allows RepartitionByExpression to pick either range or hash partitioning
-      RepartitionByExpression(distribution, query, numShufflePartitions)
+      if (Spark3VersionUtil.isSpark30) {
+        repartitionByExpressionCtor.newInstance(distribution.toSeq, query, new Integer(numShufflePartitions))
```
`new Integer` is deprecated since JDK 9, and the suggestion of replacement is `Integer.valueOf`

wypoon replied:
Thanks for the tip.

RussellSpitzer commented:
@wypoon Sorry i was on vacation for a few weeks, i'll check this out soon. I think you addressed everything I was worried about. @rdblue should take a look too but I think it would be good if we had a compatible artifact for folks to start trying out at least.

pan3793 commented:
Thanks for the great works! I have deployed iceberg master code with this patch with Spark 3.1.2, and everything goes well unit now.

pan3793 commented on spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/DistributionAndOrderingUtils.scala:
```diff
@@ -80,7 +94,11 @@ object DistributionAndOrderingUtils {
+      if (Spark3VersionUtil.isSpark30) {
+        repartitionByExpressionCtor.newInstance(distribution.toSeq, query, new Integer(numShufflePartitions))
+      } else {
+        repartitionByExpressionCtor.newInstance(distribution.toSeq, query, Some(numShufflePartitions))
```
One thing that may out of the PR's scope. There is a feature introduced in SPARK-32056, iff `numShufflePartitions` is absent, partition coalesces for AQE will effect, which will avoid writing small files when AQE is enabled.

wypoon replied:
Indeed, SPARK-32056, introduced in Spark 3.1, is what changed the signature so that the last parameter is an `Option[Int]` instead of an `Int`. If you look at the code change for SPARK-32056, if the `Option[Int]` is `None`, then the number of partitions is obtained from the conf:
```
  val numPartitions = optNumPartitions.getOrElse(SQLConf.get.numShufflePartitions)
```
And here, in `DistributionAndOrderingUtils.prepareQuery`, `numShufflePartitions` is indeed `conf.numShufflePartitions` from the `SQLConf` that is passed in.
So there is no reason to change the code here.

pan3793 replied:
Things are more interesting here, the `Strategy` also changed in SPARK-32056.
https://github.com/apache/spark/pull/28900/files#diff-21f071d73070b8257ad76e6e16ec5ed38a13d1278fe94bd42546c258a69f4410R687-R690

wypoon replied:
I can see that there will be opportunities to vary the logic depending on Spark 3 version to take advantage of new Spark 3 features. However, as you remarked, that is out of the scope of this PR. Here we just want to support building and running against both 3.0 and 3.1 and keep the logic the same.

pan3793 replied:
I'm also +1 for keeping the logic the same in this PR, just wanna know if it will introduce side effects when set `numShufflePartitions` to `None`. @RussellSpitzer @aokolnychyi would you please take a look at it?

RussellSpitzer replied:
I'm +1, I think as a follow up we should be checking if we are in Spark3 and doing something like
```
val numPartitions = write.requiredNumPartitions()
        val finalNumPartitions = if (numPartitions > 0) {
          numPartitions
        } else {
          conf.numShufflePartitions
        }
```

Which is what we are doing, but I think that's a bit of a perf improvement and not neccessary for getting this compatibility in. Maybe @wypoon you want to make a followup?

wypoon replied:
@RussellSpitzer, what is the `write` in your code example?
I can look into a followup.

RussellSpitzer replied:
Ah sorry we pass through the whole write (instead of just the distribution and ordering) as a parameter.

RussellSpitzer replied:
@aokolnychyi is busy at the moment but I believe he'll want to put in some of the code changes to make this easier

RussellSpitzer commented:
I'll do some quick tests myself but I'm good to go with this, I have one tiny additionaly suggestion that we put 3.0.1 and 3.1.1 in constants so we can change around versions a bit easier. Just have a Spark30Version and Spark31Version constant?

wypoon commented:
@RussellSpitzer thanks for looking at this again. I have just updated the PR with your suggestion to put the Spark 3.0 and 3.1 versions in constants. I also adopted a change suggested by @pan3793 in his review.

RussellSpitzer commented:
Ok I think everything is good to go here! Thank you so much @wypoon for dealing with all of my comments :) This is going to be a huge benefit to the Iceberg community!

RussellSpitzer commented:
Thanks to @rdblue and @pan3793 For reviewing as well!

wypoon commented:
@RussellSpitzer, @rdblue and @pan3793, thanks for your reviews. @RussellSpitzer especially, thanks for your help and getting this in!
